### PR TITLE
Create DEV Helm Repository for Tractus-X

### PR DESCRIPTION
**Pull request by carslen**
This PR implements:
- a GO binary, pulling released Helm charts from GH Org repositories, build DEV Helm repository
- GH Workflows for building the GO binary and executing/building the Helm repository.
- README.md file which will be shown in repo root (→ GH page).

The Helm repository will be available at https://eclipse-tractusx.github.io/tractus-x-release/charts/dev

For now this PR is a draft. After Code Review I will remove the draft flag from this PR.

## Go binary helmRepoIndex

- get a list of all GH Org repositories
- for each repository:
  - check if GH Pages are enabled, if not continue
  - If a `index.yaml` in repos GH Page is available download
  - merge downloaded  `index.yaml` with local one
  - delete downloaded index file 

### Build Go - helmRepoIndex

Build the Go binary and stores it at `.github/bin`.  `.github/bin` directory is created if not presend. Go binary is pushed back to repository to make it available for the second workflow.

### Build Helm Repository - DEV

This workflows fetches from all GH Org repositories their local Helm repository file (`index.yaml`) and merges it into the central `charts/dev/index.yaml`. No released Helm charts are fetched and stored localy.

Updated `charts/dev/index.yaml` is pushed back to repository.

## Things to discuss/ ToDo's
- Is it best practice to build a binary and push it back to the repo and use this binary with a workflow. If not, does it make sense to create a separate Git repo for this binary (and maybe upcoming other helper binaries) and to release them there? This would make the bins reusable.
- ToDo: Authenticated github client isn't nesseccary for workflow execution. This was integrated during local testing and hitting the GH API rate limit. Either remove, or add argument to Go binary to use it or not.

### CODE REVIEW

File: charts/dev/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Tractus-X - Helm Repository: DEV
+Helm charts to deploy Tractus-X products.
```
Comment (Siegfriedk) [inline charts/dev/README.md:2]: I would add a remark why its DEV; Whats the use case, whats the filter differentiation

File: charts/dev/README.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# Tractus-X - Helm Repository: DEV
```
Comment (Siegfriedk) [inline charts/dev/README.md:1]: Perhaps Tractus-x - Dev helm repository?

File: src/helmRepoIndex/main.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+	return github.NewClient(tc)
+}
+
+func getOrgRepos(ctx context.Context, gitOwner string, client *github.Client) ([]*github.Repository, *github.Response, error) {
+	opt := &github.RepositoryListByOrgOptions{
+		Type: "all",
+	}
+	repos, response, err := client.Repositories.ListByOrg(ctx, gitOwner, opt)
+	return repos, response, err
```
Comment (Siegfriedk) [inline src/helmRepoIndex/main.go:80]: Please add a new line before a return

File: src/helmRepoIndex/main.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	tc := oauth2.NewClient(ctx, ts)
+
+	return github.NewClient(tc)
+}
+
+func getOrgRepos(ctx context.Context, gitOwner string, client *github.Client) ([]*github.Repository, *github.Response, error) {
+	opt := &github.RepositoryListByOrgOptions{
+		Type: "all",
+	}
+	repos, response, err := client.Repositories.ListByOrg(ctx, gitOwner, opt)
```
Comment (Siegfriedk) [inline src/helmRepoIndex/main.go:79]: Rico used a github client too and had to verify that there are no 'next' results. is this caught here somehow?

File: src/helmRepoIndex/main.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	fullURLFile := *pageInfo.HTMLURL + "index.yaml"
+	fileName := gitRepo + "-index.yaml"
+
+	file, err := os.Create(fileName)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	response, _ := http.Get(fullURLFile)
+	if response.StatusCode == 200 {
```
Comment (Siegfriedk) [inline src/helmRepoIndex/main.go:94]: what if its not 200? We probably need more logging here

File: src/helmRepoIndex/main.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		// Close file
+		err = file.Close()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	return fileName
+}
+
+func buildHelmRepoIndex(indexFile string, mergeIndexFile string) {
```
Comment (Siegfriedk) [inline src/helmRepoIndex/main.go:112]: You are failing the whole build process if one repo produces an issue. we should make it more resiliant and skip a repository with a warning/error log if that happens

File: src/helmRepoIndex/main.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+		if err != nil {
+			log.Fatal(err)
+		}
+		repoFile.Merge(newIndex)
+		repoFile.Generated = time.Now()
+		err = repoFile.WriteFile(indexFile, 0644)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
```
Comment (Siegfriedk) [inline src/helmRepoIndex/main.go:130]: Please add a new line here

File: src/helmRepoIndex/main.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// merge index only if file has content and is not from local repository
+	if fileStat, _ := os.Stat(mergeIndexFile); fileStat.Size() > 0 && mergeIndexFile != "tractusx-release" {
+		newIndex, err := repo.LoadIndexFile(mergeIndexFile)
+		if err != nil {
+			log.Fatal(err)
+		}
```
Comment (Siegfriedk) [inline src/helmRepoIndex/main.go:123]: please add a new line here

**Review by Siegfriedk — APPROVED** [review]

File: src/helmRepoIndex/main.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+	"os"
+	"time"
+
+	"github.com/google/go-github/v48/github"
+	"golang.org/x/oauth2"
+	"helm.sh/helm/v3/pkg/repo"
+)
+
+func main() {
+	var centralHelmIndex string
```
Comment (sTb1337) [inline src/helmRepoIndex/main.go:39]: How will this be automatically tested? It would be good to have some unit tests.

**Comment by carslen** [comment]
As per Tractus-X/Catena-X decision the Helm repositories should be managed/located in a different repository than _tractus-x-release_. Close this PR due to this decision.

File: src/helmRepoIndex/main.go (hunk 4 above)
Comment (carslen) [reply src/helmRepoIndex/main.go:79]: As `PerPage` is not explicit defined for `github.ListOptions` (which defaults to 0 to my understanding), I expect no issues here. [Package documentation](https://pkg.go.dev/github.com/google/go-github/v48@v48.1.0/github#hdr-Pagination) also state, that it is supported. I don't derive a _you must conifigure it_ from doc.

File: src/helmRepoIndex/main.go (hunk 5 above)
Comment (carslen) [reply src/helmRepoIndex/main.go:94]: I'll investigate on that...

File: src/helmRepoIndex/main.go (hunk 6 above)
Comment (carslen) [reply src/helmRepoIndex/main.go:112]: Hm, why? func buildHelmRepoIndex is executed when repository has a valid GH pages setup but if GH Page delivers no `index.yaml` no merge will be executed.

Do you have another issue in mind? Ok, currently we rely on a consistent downloaded `index.yaml` if there is one.

File: src/helmRepoIndex/main.go (hunk 4 above)
Comment (Siegfriedk) [reply src/helmRepoIndex/main.go:79]: i'm not concerned that we need to but that the amount of results you get is limited if you don't fetch the rest of it.

**Pull request closed without merging**